File: testdata/fixtures/clojure/CWE-502.clj
;; CWE-502: Deserialization of Untrusted Data via read-string
;; read-string without :exclude can evaluate arbitrary code

(ns fixture.cwe502)

(defn vuln-read [x]
  ;; ruleid: read-string-unsafe
  (read-string x)
  x)

(defn vuln-read2 [y]
  ;; ruleid: read-string-unsafe
  (read-string y)
  y)

(defn vuln-read3 [z]
  ;; ruleid: read-string-unsafe
  (read-string z)
  z)

(defn vuln-read4 [a]
  ;; ruleid: read-string-unsafe
  (read-string a)
  a)

;; ok: read-string-unsafe
(defn safe-outside [z]
  (str z))

;; ok: read-string-unsafe
(defn safe-plain [s]
  (pr-str s))

;; ok: read-string-unsafe
(defn safe-pr [s]
  (pr-str s))

File: testdata/fixtures/rust/CWE-502.rs
// CWE-502: Deserialization
fn vuln(data: &str) {
    // ruleid: rust-deserialization
    let _: String = bincode::deserialize(data).unwrap();
    // ruleid: rust-deserialization
    let _: String = serde_json::from_str(data).unwrap();
    // ok: rust-deserialization
    let _: String = bincode::deserialize("fixed").unwrap();
    // ok: rust-deserialization
    let _: String = serde_json::from_str("{\"a\":1}").unwrap();
}
